### DISCUSSION TITLE
Issues with CLI installation on windows 7

### DISCUSSION THREAD

**Discussion by firstmiddlelast** [jscad/OpenJSCAD.org#1076] [General]
Hi! I'm stuck with windows 7, so my Node.Js version is 13.6.0 (don't tell me to upgrade please) ; I have npm 6.13.4, and the `npm i  @jscad/cli` command ends correctly, but with the message : 
```
npm WARN saveError ENOENT: no such file or directory, open 'C:\Users\ROBIN\dev\openjscad\package.json'
npm WARN enoent ENOENT: no such file or directory, open 'C:\Users\ROBIN\dev\openjscad\package.json'. 

```
But when I run `jscad`, this happens :
 
```
internal/modules/cjs/loader.js:628
  throw e;
  ^
Error: No valid exports main found for 'C:\Users\ROBIN\AppData\Roaming\npm\node_modules\@jscad\cli\node_modules\fflate'
←[90m    at resolveExportsTarget (internal/modules/cjs/loader.js:625:9)←[39m

[...]

←[90m    at require (internal/modules/cjs/helpers.js:72:18)←[39m
    at Object.<anonymous> (C:\Users\ROBIN\AppData\Roaming\npm\node_modules\←[4m@jscad←[24m\cli\node_modules\←[4m@jscad←[
24m\3mf-serializer\src\index.js:34:17)
←[90m    at Module._compile (internal/modules/cjs/loader.js:1144:30)←[39m {
  code: ←[32m'MODULE_NOT_FOUND'←[39m
```

..obviously, the `3mf-serializer\src\index.js` line : 
`const zipSync = require('fflate').zipSync`
can't load the fflate library. 
I have tried installing fflate separately, copy/pasting the fflate directory into the node_modules.. nothing changes : fflate can't be loaded. 
Does anyone have any idea what could be done? Downgrade npm or fflate to a previous version? Some dirty hack would be ok, too, or even better, using another javascript engine than node.js (but my hopes on this are low :). 
Oh, and as for my use case : I would like to use the CLI as an external modeling / generative tool, with such workflows as : 
* my app mesh => stl export => jscad operations on those files => stl import into my app
* my app generative algorithm => jscad code generation => stl import into my app

**Comment by z3dev** [comment]
@firstmiddlelast welcome to the world of never ending changes.

We (JSCAD) test and support the latest versions of NODEJS and NPM. Sorry.

having said that, there were working versions of JSCAD that used those versions of NODEJS and NPM. And there was a know issue on Windows which prevented the installation, but there was a work around. Have you tried any old versions? You will have to edit package.json, and change the versions of 'dependencies', and 'dev-dependencies'

**Reply by z3dev** [reply]
See issue #765

**Reply by firstmiddlelast** [reply]
Quick update, hoping this may be useful to someone someday : I uninstalled everything and deleted all directories manually, then installed nvm for windows - this didn't work the first time I tried because I wanted to install it in `program files`, but this failed (maybe because of spaces in the name), so I let it install itself into the directory the installer wanted (`appdata\roaming`, yuck) , then opened a command prompt as an administrator (this is mandatory), and did
```
nvm install 12.16.3 32
nvm use 12.16.3 32
npm install @jscad/cli
```

(`32` is mandatory, nvm doesn't properly detect this is Windows 7), and _then_ I could use `jscad `in the command line. 
Welcome to the wonderful world where you have to be an administrator to install an installer that installs an environment in order to install an installer that lets you run... javascript files. 
Anyway, if things do end up working, I hope I can post happy rendered screenshots. 

EDIT : installing the examples with 
```
npm install @jscad/examples
```
then moving the `node_modules/@jscad/examples` directory into the `.` directory for easy access was a great help. 
Actually, CLI script syntax is not exactly the same as the one for the Web, as described at [https://en.wikibooks.org/wiki/OpenJSCAD_User_Guide#Anatomy_of_a_OpenJSCAD_Script](https://en.wikibooks.org/wiki/OpenJSCAD_User_Guide#Anatomy_of_a_OpenJSCAD_Script) . In CLI scripts, the JSCAD files have to be `require`'d and the `main` function has to be exported. 
Example : 
```
const { sphere} = require('@jscad/modeling').primitives

function main () {
   return sphere();
}

module.exports = {main}
```
...In the end, this is working as expected. Now let's get on to some productive work!

**Reply by z3dev** [reply]
@firstmiddlelast that's cool. And thanks for the excellent information.

Yes. I meant installing specific versions of NODEJS and NPM. I think everything JSCAD should work, however there are dependencies on other libraries, which may issues.

**Comment by firstmiddlelast** [comment]
Oh, you mean old versions of JSCAD? I can try that. Any suggestion on which version I should try? 
Thanks for your answer!
I'm certainly not new to the world of never ending changes, but to Node.js and npm, I am :).
